src/flint/types/fmpz_mod.pyx

```diff
@@ -0,0 +1,364 @@
+        # Ensure modulus is fmpz type
+        if not typecheck(mod, fmpz):
+            mod = any_as_fmpz(mod)
+            if mod is NotImplemented:
+                raise TypeError("Context modulus must be able to be case to an `fmpz` type")
+
+        # Ensure modulus is positive
+        if mod < 1:
+            raise ValueError("Modulus is expected to be positive")
+
+        # Init the context
+        fmpz_mod_ctx_init(self.val, (<fmpz>mod).val)
+    
+    def modulus(self):
+        """
+        Return the modulus from the context as an fmpz
+        type
+
+            >>> mod_ctx = fmpz_mod_ctx(2**127 - 1)
+            >>> mod_ctx.modulus()
+            170141183460469231731687303715884105727
+
+        """
+        n = fmpz()
+        fmpz_set(n.val, (<fmpz_t>self.val.n))
+        return n
+
+    def __eq__(self, other):
+        if typecheck(other, fmpz_mod_ctx):
+            return fmpz_equal(self.val.n, (<fmpz_mod_ctx>other).val.n)
+        return False
+
+    def __hash__(self):
+        return hash(self.modulus())
+
+    def __str__(self):
+        return f"Context for fmpz_mod with modulus: {self.modulus()}"
+
+    def __repr__(self):
+        return f"fmpz_mod_ctx({self.modulus()})"
+
+    def __call__(self, val):
+        return fmpz_mod(val, self)
+
+cdef class fmpz_mod(flint_scalar):
+    """
+    The *fmpz_mod* type represents integer modulo an 
+    arbitrary-size modulus. For wordsize modulus, see
+    :class:`~.nmod`.
+
+    An *fmpz_mod* element is constructed from an :class:`~.fmpz_mod_ctx`
+    either by passing it as an argument to the type, or
+    by directly calling the context
+
+        >>> fmpz_mod(-1, fmpz_mod_ctx(2**127 - 1))
+        fmpz_mod(170141183460469231731687303715884105726, 170141183460469231731687303715884105727)
+        >>> ZmodN = fmpz_mod_ctx(2**127 - 1)
+        >>> ZmodN(-2)
+        fmpz_mod(170141183460469231731687303715884105725, 170141183460469231731687303715884105727)
+
+    """
+
+    def __cinit__(self):
+        fmpz_init(self.val)
+
+    def __dealloc__(self):
+        fmpz_clear(self.val)
+
+    def __init__(self, val, ctx):
+        if not typecheck(ctx, fmpz_mod_ctx):
+            raise TypeError
+        self.ctx = ctx
+
+        # When the input is also an fmpz_mod we just need
+        # moduli to match
+        if typecheck(val, fmpz_mod):
+            if self.ctx != (<fmpz_mod>val).ctx:
+                raise ValueError("moduli must match")
+            # fmpz_mod_set_fmpz(self.val, (<fmpz_mod>val).val, self.ctx.val)
+            fmpz_set(self.val, (<fmpz_mod>val).val)
+            return
+
+        # For all other cases, the easiest is to first convert to
+        # fmpz type and set this way
+        if not typecheck(val, fmpz):
+            val = any_as_fmpz(val)
+            if val is NotImplemented:
+                raise NotImplementedError
+        fmpz_mod_set_fmpz(self.val, (<fmpz>val).val, self.ctx.val)
+
+    def any_as_fmpz_mod(self, obj):
```

Comment:
Can this be `cdef`?

Reply:
Or otherwise a leading underscore like `_any_as...`

Reply:
currently it shows in the docs but it should not be there.

Reply:
Made it a cdef